shesmu-pluginapi/src/main/java/ca/on/oicr/gsi/shesmu/plugin/filter/FilterBuilder.java

```diff
@@ -1,18 +1,148 @@
+          final FilterAdded result = new FilterAdded();
+          result.setStart(start.map(Instant::getEpochSecond).orElse(null));
+          result.setEnd(end.map(Instant::getEpochSecond).orElse(null));
+          return result;
+        }
 
-  F negate(F filter);
+        @Override
+        public FilterJson addedAgo(long offset) {
+          final FilterAddedAgo result = new FilterAddedAgo();
+          result.setOffset(offset);
+          return result;
+        }
+
+        @Override
+        public FilterJson and(Stream<FilterJson> filters) {
+          final FilterAnd result = new FilterAnd();
+          result.setFilters(filters.toArray(FilterJson[]::new));
+          return result;
+        }
+
+        @Override
+        public FilterJson checked(Optional<Instant> start, Optional<Instant> end) {
+          final FilterChecked result = new FilterChecked();
+          result.setStart(start.map(Instant::getEpochSecond).orElse(null));
+          result.setEnd(end.map(Instant::getEpochSecond).orElse(null));
+          return null;
+        }
+
+        @Override
+        public FilterJson checkedAgo(long offset) {
+          final FilterCheckedAgo result = new FilterCheckedAgo();
+          result.setOffset(offset);
+          return result;
+        }
+
+        @Override
+        public FilterJson external(Optional<Instant> start, Optional<Instant> end) {
+          final FilterExternal result = new FilterExternal();
+          result.setStart(start.map(Instant::getEpochSecond).orElse(null));
+          result.setEnd(end.map(Instant::getEpochSecond).orElse(null));
+          return null;
+        }
+
+        @Override
+        public FilterJson externalAgo(long offset) {
+          final FilterExternalAgo result = new FilterExternalAgo();
+          result.setOffset(offset);
+          return result;
+        }
+
+        @Override
+        public FilterJson fromFile(String... files) {
+          final FilterSourceFile result = new FilterSourceFile();
+          result.setFiles(files);
+          return result;
+        }
+
+        @Override
+        public FilterJson fromSourceLocation(Stream<LocationJson> locations) {
+          final FilterSourceLocation result = new FilterSourceLocation();
+          result.setLocations(locations.map(LocationJson::new).toArray(LocationJson[]::new));
+          return result;
+        }
 
+        @Override
+        public FilterJson ids(List<String> ids) {
+          final FilterIds result = new FilterIds();
+          result.setIds(ids);
+          return result;
+        }
+
+        @Override
+        public FilterJson isState(ActionState... states) {
+          final FilterStatus result = new FilterStatus();
+          result.setState(states);
+          return result;
+        }
+
+        @Override
+        public FilterJson negate(FilterJson filter) {
+          final FilterJson copy = filter.convert(this);
+          copy.setNegate(!copy.isNegate());
```

Comment:
As per our discussion: Thanks, I hate it. Please document :)

Reply:
Actually, it makes a copy, so it's safer than I remember. Hooray, (it/I) is not a total hack.